src/map.js

```diff
@@ -1571,14 +1597,53 @@ var map = function (arg) {
    *        undefined, use the default (white).  Otherwise, a css color or
    *        CanvasRenderingContext2D.fillStyle to fill the initial canvas.
    *        This could match the background of the browser page, for instance.
-   * @returns {string|HTMLCanvasElement}: data URL with the result or the
-   *    HTMLCanvasElement with the result.
+   *    wait: if 'idle', wait for the map to be idle and one animation frame to
+   *        occur.  If truthy, wait for an animation frame to occur.
+   *        Otherwise, take the screenshot as sson as possible.
+   *    attribution: if null or unspecified, include the attribution only if
+   *        all layers are used.  If false, never include the attribution.  If
+   *        true, always include it.
+   * @returns {deferred}: a jQuery Deferred object.  The done function receives
+   *    either a data URL or the HTMLCanvasElement with the result.
    */
   this.screenshot = function (layers, type, encoderOptions, opts) {
+    var defer;
+
+    if (layers && !Array.isArray(layers) && !layers.renderer) {
+      type = type || layers.type;
+      encoderOptions = encoderOptions || layers.encoderOptions;
+      opts = opts || layers;
+      layers = layers.layers;
+    }
     opts = opts || {};
+    /* if asked to wait, return a Deferred that will do so, calling the
+     * screenshow function without waiting once it is done. */
+    if (opts.wait) {
+      var optsWithoutWait = $.extend({}, opts, {wait: false});
+      defer = $.Deferred();
+
+      var waitForRAF = function () {
+        window.requestAnimationFrame(function () {
+          defer.resolve();
+        });
+      };
+
+      if (opts.wait === 'idle') {
+        m_this.onIdle(waitForRAF);
+      } else {
+        waitForRAF();
+      }
+      return defer.then(function () {
+        return m_this.screenshot(layers, type, encoderOptions, optsWithoutWait);
+      });
+    }
+    defer = $.when();
```

Comment:
do we ever return this resolved promise (since after this line we re-assign defer)?